src/java/io/bazel/rulesscala/worker/GenericWorker.java

```diff
@@ -1,117 +0,0 @@
-package io.bazel.rulesscala.worker;
-
-import static java.nio.charset.StandardCharsets.UTF_8;
-
-import com.google.devtools.build.lib.worker.WorkerProtocol.WorkRequest;
-import com.google.devtools.build.lib.worker.WorkerProtocol.WorkResponse;
-import java.io.ByteArrayOutputStream;
-import java.io.IOException;
-import java.io.PrintStream;
-import java.nio.file.Files;
-import java.nio.file.Paths;
-import java.util.Arrays;
-import java.util.List;
-
-public class GenericWorker {
-  protected final Processor processor;
-
-  public GenericWorker(Processor p) {
-    processor = p;
-  }
-
-  protected void setupOutput(PrintStream ps) {
-    System.setOut(ps);
-    System.setErr(ps);
-  }
-
-  // Mostly lifted from bazel
-  private void runPersistentWorker() throws IOException {
-    PrintStream originalStdOut = System.out;
-    PrintStream originalStdErr = System.err;
-
-    while (true) {
-      try {
-        WorkRequest request = WorkRequest.parseDelimitedFrom(System.in);
-        if (request == null) {
-          break;
-        }
-        ByteArrayOutputStream baos = new ByteArrayOutputStream();
-        int exitCode = 0;
-
-        try (PrintStream ps = new PrintStream(baos)) {
-          setupOutput(ps);
```

Comment:
They create a new `ByteArrayOutputStream` for every message and then use "try with resources" and you just reset the `ByteArrayOutputStream` for every message? If so any reason why you preferred your approach?

Reply:
[Resetting a `ByteArrayOutputStream`](https://docs.oracle.com/javase/7/docs/api/java/io/ByteArrayOutputStream.html#reset()) allows us to reuse existing memory instead of allocating a new internal buffer, so it should be a tiny bit more efficient.

Reply:
Now you have me worried. Resetting the buffer, as far as I can tell, will never shrink the internal byte array-- so it will grow to the max output of any previous work request.

There's probably a fun micro optimization here, but the safe play is to just recreate the buffer on each run.

Reply:
I'm taking a middle ground here and subclassing `ByteArrayOutputStream` so that it only resets the internal buffer if it's above a default size threshold.